README.md
# Logstash KS3 Output Plugin

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## 文档 Documentation

该插件可以累积上传 logstash events 到 KS3

首先，你需要拥有 KS3 的权限 (ak/sk) 和一个可写的bucket

上传到 KS3 前，该插件会将文件写到临时目录

_This plugin batches and uploads logstash events into Ksyun Object Storage Service (Ksyun KS3)._

_First, you should have a writable bucket and ks3 access permissions((Typically access_key_id and access_key_secret))._

_ks3 output plugin creates temporary files into the OS' temporary directory(You can set this configuration by **temporary_directory** option) before uploading them to ks3._


临时文件路径如下：

_ks3 output plugin output files have the following format_

```bash
/tmp/logstash/ks3/eaced620-e972-0136-2a14-02b7449ba0a9/logstash/1/ls.ks3.e27ff60b-98eb-42f8-87bb-09cdb56102c2.2021-08-03T20.16.part-0.data
```

|||
|---|---|
|/tmp/logstash/ks3| **temporary_directory** 选项指定的临时目录，默认为系统临时目录 |
|eaced620-e972-0136-2a14-02b7449ba0a9 | 随机 uuid |
|logstash/1|ks3 对象前缀|
|ls.ks3|标明是 logstash 插件|
|eaced620-e972-0136-2a14-02b7449ba0a9 | 随机 uuid |
|2018-12-24T14.27 | 创建时间 |
|part-0|分块|
|.data|后缀, `encoding` 是 gzip 的话，结尾为 .gz ，其他是 .data|

### 使用方法
样例:
```ruby
input {
  file {
    path => "/etc/logstash-7.3.0/sample.data"
    codec => json {
      charset => "UTF-8"
    }
  }
}

output {
  ks3 {
    endpoint => "ks3 endpoint"                            (required)
    bucket => "bucket"                                    (required)
    access_key_id => "ak"                                 (required)
    access_key_secret => "sk"                             (required)
    prefix => "logstash/ks3"                              (optional, default = "")
    recover => true                                       (optional, default = true)
    rotation_strategy => "size_and_time"                  (optional, default = "size_and_time")
    time_rotate => 15                                     (optional, default = 15) - Minutes
    size_rotate => 31457280                               (optional, default = 31457280) - Bytes
    encoding => "gzip"                                    (optional, default = "none")
    additional_ks3_settings => {                          (optional, default = 1024)
      secure_connection_enabled => false                  (optional, default = false)
      server_side_encryption_algorithm => "AES256"        (optional, default = "none")
    }
    codec => json {
      charset => "UTF-8"
    }
  }
}
```

### 插件配置文件选项

|Configuration|Type|Required|Comments|
|:---:|:---:|:---:|:---|
|endpoint|string|Yes|endpoint|
|bucket|string|Yes|bucket|
|access_key_id|string|Yes|ak|
|access_key_secret|string|Yes|sk|
|prefix|string|No|前缀|
|recover|string|No|插件支持崩溃恢复crash recovery, 可以设置 **recover** 为true，来从异常崩溃中恢复上传。|
|additional_ks3_settings|hash|No|附加的客户端配置: `server_side_encryption_algorithm`（服务端加密算法）, `secure_connection_enabled` （是否安全连接https/http）|
|temporary_directory|string|No|指定的临时目录，默认为系统临时目录|
|rotation_strategy|string|No|文件滚动更新策略。可选值：size、time、size_and_time（默认）|
|size_rotate|number|No|如果文件大小大于等于size_rotate，将滚动更新文件（依赖rotation_strategy）。默认为30 MBytes|
|time_rotate|number|No|如果文件的生存时长大于等于time_rotate，将滚动更新文件（依赖rotation_strategy）。默认为15分钟|
|upload_workers_count|number|No|上传线程并发数|
|upload_queue_size|number|No|上传队列大小|
|encoding|string|No|是否启用gzip压缩 `gzip` and `none`|

## 部署

```bash
./bin/logstash-plugin install /logstash-output-ks3/logstash-output-ks3-0.0.1-java.gem
```
结果如下：

```bash
Validating logstash-output-ks3
Installing logstash-output-ks3
Installation successful
```

查看插件 :
```bash
./bin/logstash-plugin list --verbose logstash-output-ks3

logstash-output-ks3 (0.0.1)
```

## 开发

### 1. 插件开发测试

#### Code
- 开始前需要JRuby

- 安装依赖
```sh
bundle install
```

### 2. 在 Logstash 运行你的插件

#### 2.1 编译、安装、运行

- 编译

```sh
gem build logstash-output-ks3.gemspec
```

- 安装到logstash

```sh
bin/logstash-plugin install /path/to/logstash-output-ks3-0.1.1-java.gem
```

- 在logstash测试

```bash
./bin/logstash -f config/logstash-sample.conf
```

## 引用
[logstash-output-oss](https://github.com/aliyun/logstash-output-oss)


CHANGELOG.md
## 0.0.1
  - init

lib/logstash/outputs/ks3.rb
# encoding: utf-8
require "logstash/outputs/base"
require 'concurrent'
require 'logstash-output-ks3_jars'

java_import "com.ksyun.ks3.service.Ks3ClientConfig";
java_import "com.ksyun.ks3.http.HttpClientConfig";
java_import "com.ksyun.ks3.service.Ks3Client";


#
#  Logstash KS3 Output Plugin
#
# Usage:
# In order to write output data to ks3, you should add configurations like below to logstash
# output {
#   ks3 {
#     "endpoint" => "ks3 endpoint to connect to"              (required)
#     "bucket" => "Your bucket name"                          (required)
#     "access_key_id" => "Your access key id"                 (required)
#     "access_key_secret" => "Your access secret key"         (required)
#     "prefix" => "logstash/%{index}"                         (optional, default = "")
#     "recover" => true                                       (optional, default = true)
#     "rotation_strategy" => "size_and_time"                  (optional, default = "size_and_time")
#     "time_rotate" => 15                                     (optional, default = 15) - Minutes
#     "size_rotate" => 31457280                               (optional, default = 31457280) - Bytes
#     "encoding" => "none"                                    (optional, default = "none")
#     "additional_ks3_settings" => {
#       "max_connections_to_ks3" => 1024                      (optional, default = 1024)
#       "secure_connection_enabled" => false                  (optional, default = false)
#     }
#   }
# }
#
class LogStash::Outputs::Ks3 < LogStash::Outputs::Base
  require 'logstash/outputs/ks3/rotations/hybrid_rotation'
  require 'logstash/outputs/ks3/file_uploader'
  require 'logstash/outputs/ks3/file_manager'
  require 'logstash/outputs/ks3/version'

  ROTATE_CHECK_INTERVAL_IN_SECONDS = 15

  MAX_CONNECTIONS_TO_KS3_KEY = "max_connections_to_ks3"

  SERVER_SIDE_ENCRYPTION_ALGORITHM_KEY = "server_side_encryption_algorithm"

  SECURE_CONNECTION_ENABLED_KEY = "secure_connection_enabled"

  config_name "ks3"

  concurrency :shared

  default :codec, "line"

  # ks3 bucket name
  config :bucket, :validate => :string, :required => true

  # ks3 endpoint to connect
  config :endpoint, :validate => :string, :required => true

  # access key id
  config :access_key_id, :validate => :string, :required => true

  # access secret key
  config :access_key_secret, :validate => :string, :required => true

  # additional ks3 client configurations, valid keys are:
  # server_side_encryption_algorithm(server side encryption, only support AES256 now)
  # secure_connection_enabled(enable https or not)
  # max_connections_to_ks3(max connections to ks3)
  # TODO: add other ks3 configurations
  config :additional_ks3_settings, :validate => :hash, :required => false

  # rotate this file if its size greater or equal than `size_rotate`
  config :size_rotate, :validate => :number, :default => 30 * 1024 * 1024

  # rotate this file if its life time greater or equal than `time_rotate`
  config :time_rotate, :validate => :number, :default => 15

  # if true, ks3 plugin will recover files since last crash
  config :recover, :validate => :boolean, :default => true

  # temporary directory that used to cache event before upload to ks3
  config :temporary_directory, :validate => :string, :default => File.join(Dir.tmpdir, "logstash/ks3")

  # prefix that added to generated file name
  # sample file name looks like
  # `prefix`/logstash.ks3.{random-uuid}.{%Y-%m-%dT%H.%M}.part-{index}.{extension}
  # WARNING: this option support string interpolation, so it may create a lot of temporary local files
  config :prefix, :validate => :string, :default => ''

  # file rotation strategy
  config :rotation_strategy, :validate => %w(size time size_and_time), :default => "size_and_time"

  # concurrent number of upload threads
  config :upload_workers_count, :validate => :number, :default => (Concurrent.processor_count * 0.5).ceil

  # upload queue size
  config :upload_queue_size, :validate => :number, :default => 2 * (Concurrent.processor_count * 0.25).ceil

  # support plain and gzip compression before upload to ks3
  config :encoding, :validate => %w(none gzip), :default => "none"

  public
  def register
    # check if temporary_directory is writable
    begin
      FileUtils.mkdir_p(@temporary_directory) unless Dir.exist?(@temporary_directory)
      ::File.writable?(@temporary_directory)
    rescue
      raise LogStash::ConfigurationError, "Logstash KS3 Output Plugin can not write data to " + @temporary_directory
    end

    # check rotation configuration
    if @size_rotate.nil? and @time_rotate.nil? || @size_rotate <= 0 && @time_rotate <= 0
      raise LogStash::ConfigurationError, "Logstash KS3 Output Plugin must have at least one of time_file or size_file set to a value greater than 0"
    end

    if @upload_workers_count <= 0 || @upload_queue_size <= 0
      raise LogStash::ConfigurationError,  "Logstash KS3 Output Plugin must have both upload_workers_count and upload_queue_size are positive"
    end

    # create upload thread pool
    executor = Concurrent::ThreadPoolExecutor.new({ :min_threads => 1,
                                                    :max_threads => @upload_workers_count,
                                                    :max_queue => @upload_queue_size,
                                                    :fallback_policy => :caller_runs })

    # get file rotation strategy
    @rotation = rotation

    # initialize ks3 client
    @ks3 = initialize_ks3_client

    # initialize file uploader
    @file_uploader = FileUploader.new(@ks3, @bucket, @additional_ks3_settings, @logger, executor)

    # initialize file manager
    @file_manager = FileManager.new(@logger, @encoding, @temporary_directory)

    # recover from crash
    recover_from_crash if @recover

    # start rotate check
    start_rotate_check if @rotation.needs_periodic_check?
  end # def register

  public
  def multi_receive_encoded(events_and_encoded)
    prefixes = Set.new
    events_and_encoded.each do |event, encoded|
      prefix = event.sprintf(@prefix)
      prefixes << prefix

      begin
        @file_manager.get_file_generator(prefix) { |generator| generator.current_file.write(encoded) }
      rescue Errno::ENOSPC => e
        @logger.error("Logstash KS3 Output Plugin: No space left in temporary directory", :temporary_directory => @temporary_directory)
        raise e
      end
    end
    rotate(prefixes)
  end

  def close
    @logger.info("Logstash KS3 Output Plugin is shutting down...")

    # stop rotate check
    stop_rotate_check if @rotation.needs_periodic_check?

    prefixes = @file_manager.prefixes
    prefixes.each do |prefix|
      @file_manager.get_file_generator(prefix) do |generator|
        file = generator.current_file
        file.close
        if file.size > 0
          # upload async
          @file_uploader.upload_async(file, :on_complete => method(:clean_temporary_file))
        end
      end
    end

    @file_manager.close

    # stop file uploader
    @file_uploader.close
  end

  private
  def initialize_ks3_client
    clientConfig = Ks3ClientConfig.new()
    clientConfig.setEndpoint(@endpoint)

    hconfig = HttpClientConfig.new()
    clientConfig.setHttpClientConfig(hconfig)

    unless @additional_ks3_settings.nil?
      if @additional_ks3_settings.include?(SECURE_CONNECTION_ENABLED_KEY)
        clientConfig.setProtocol(@additional_ks3_settings[SECURE_CONNECTION_ENABLED_KEY] ?
          Ks3ClientConfig::PROTOCOL::https : Ks3ClientConfig::PROTOCOL::http)
      end
    end

    @logger.info("bucket", :bucket => @bucket)
    Ks3Client.new(@access_key_id, @access_key_secret, clientConfig)
  end

  private
  def rotation
    case @rotation_strategy
    when "size"
      SizeBasedRotation.new(size_rotate)
    when "time"
      TimeBasedRotation.new(time_rotate)
    when "size_and_time"
      HybridRotation.new(size_rotate, time_rotate)
    end
  end

  private
  def recover_from_crash
    @logger.info("Logstash KS3 Output Plugin starts to recover from crash and uploading...")
    Dir.glob(::File.join(@temporary_directory, "**/*"))
      .select {|file| ::File.file?(file) }
      .each do |file|
        temporary_file = TemporaryFile.create_existing_file(file, @temporary_directory)
        if temporary_file.size > 0
          @file_uploader.upload_async(temporary_file, :on_complete => method(:clean_temporary_file))
        else
          clean_temporary_file(temporary_file)
        end
      end
  end

  private
  def rotate(prefixes)
    prefixes.each do |prefix|
      @file_manager.get_file_generator(prefix) do |file_generator|
        file = file_generator.current_file
        if @rotation.rotate?(file)
          @logger.info("Logstash KS3 Output Plugin starts to rotate file",
                       :strategy => @rotation.class.name,
                       :key => file.key,
                       :path => file.path,
                       :size => file.size,
                       :thread => Thread.current.to_s)
          file.close
          if file.size > 0
            # upload async
            @file_uploader.upload_async(file, :on_complete => method(:clean_temporary_file))
          end
          file_generator.rotate
        end
      end
    end
  end

  private
  def clean_temporary_file(file)
    @logger.debug("Logstash KS3 Output Plugin: starts to remove temporary file",
                 :file => file.path)
    file.delete!
  end

  private
  def start_rotate_check
    @rotate_check = Concurrent::TimerTask.new(:execution_interval => ROTATE_CHECK_INTERVAL_IN_SECONDS) do
      @logger.debug("Logstash KS3 Output Plugin: starts rotation check")

      rotate(@file_manager.prefixes)
    end

    @rotate_check.execute
  end

  private
  def stop_rotate_check
    @rotate_check.shutdown
  end
end # class LogStash::Outputs::Ks3